github/github.go

```diff
@@ -78,6 +95,14 @@ func (c *Config) NewClient() (*Client, error) {
 	} else {
 		client = github.NewClient(httpClient)
 		githubBaseURL = "https://github.com/"
+		if len(c.ProxyUrl) > 0 {
+			proxyUrl, err := url.Parse(c.ProxyUrl)
+			if err != nil {
+				return nil, fmt.Errorf("proxy client creation failed: %v", err)
+			}
+			client.BaseURL = proxyUrl
+			client.UploadURL = proxyUrl
```

Comment:
I believe the upload URL is hightly dependent on the environment.

The default base URL is `https://api.github.com/` and the upload URL is `https://uploads.github.com/` for GitHub Cloud. In GHES it's like `$GITHUB_ENDPOINT/api/uploads` or `api.$GITHUB_ENDPOINT` depending on the situation.

That said, would you mind adding another flag like `github-proxy-upload-url` for a more fine-grained configuration? The default value for that would be `github-proxy-url` + `/api/uploads/`.

Reply:
Also, `githubBaserURL` should be assigned the same value as  `proxyURL` in this case, so that it is correctly propagated down to runners via `GITHUB_URL` env.

Reply:
> That said, would you mind adding another flag like github-proxy-upload-url for a more fine-grained configuration?

good call - I'll add that - thanks

> Also, githubBaserURL should be assigned the same value as proxyURL in this case, so that it is correctly propagated down to runners via GITHUB_URL env.

This one I "intentionally" kept it as github.com as (in our example/infra) we still want runners to be registered with github.com (or we didn't had the need to have a custom proxy API for runners registration nor custom internal API calls triggered by the runner). What about having another flag `github-proxy-runner-url`  to let people chose if they want to use their custom proxy URL, github.com or whatever URL they would like to, they can configure it as needed, and probably default to proxy url if not defined/empty

Reply:
let me know what do you think based on 7a976cc17b661a4d7c49e24a8c37549c752c5cb5

Reply:
Thanks. Then, I'd name it `runner-github-url` so that it sounds more like it's basically `GITHUB_URL` only for runners :)

Reply:
Also, I started to think it might communicate better if we omitted the `proxy` concept out of those names.

- `github-proxy-url`- can be just `github-url`. This is slightly more consistent with the existing `github-enterprise-url` imho
- `github-proxy-upload-url`- can also be just `github-upload-url`
- `github-proxy-runner-url` would better be `runner-github-url`
- `github-proxy-username` would better be `github-basicauth-username` so that it directly matches that of our HTTP client parameter
- `github-proxy-password` would better be `github-basicauth-password`. same as above

After all the new code path will look a bit different, like

```
client = github.NewClient(httpClient)
githubBaseURL = "https://github.com/"

if len(c.GitHubURL)  > 0 {
	baseUrl, err := url.Parse(c.GitHubURL)
	if err != nil {
		return nil, fmt.Errorf("proxy client creation failed: %v", err)
	}
	if !strings.HasSuffix(baseUrl.Path, "/") {
		baseUrl.Path += "/"
	}
	client.BaseURL = baseUrl
}
```

In the flag usage description one would introduce `github-basicauth-username` and `github-basicauth-password` are used only when you want the client to use basic auth instead of PAT or GitHub App authn in case you're running your own github API proxy.

I imagine there would be a github API proxy without even basic-auth. So an option to not enforce basic-auth when you're using proxy would also be nice.

So the below code needs a bit change from my perspective:

```
if len(c.ProxyUrl) > 0 {
		transport = ProxyTransport{BaseUrl: c.ProxyUrl, Username: c.ProxyUsername, Password: c.ProxyPassword}
	} else if len(c.Token) > 0 {
```

Perhaps it would better align with my idea if it was like:

```
if c.ProxyUsername != "" && c.ProxyPassword {
		transport = ProxyTransport{BaseUrl: c.ProxyUrl, Username: c.ProxyUsername, Password: c.ProxyPassword}
	} else if len(c.Token) > 0 {
```

Sorry for the loooong reply but I believe with some tweaks your change will be perfect! I appreciate your cooperation.

Reply:
thanks for the feedback - the proposal looks good to me 👍 - I've done the renaming on 891f909dc5137b5dd5a54d10051ecfacc893b0d3